Source/ACE/Entity/Player.cs

```diff
@@ -243,20 +243,59 @@ public Player(Session session) : base(ObjectType.Creature, session.CharacterRequ
             Session.WorldSession.EnqueueSend(setTurbineChatChannels, general, trade, lfg, roleplay);
         }
 
+        /// <summary>
+        /// Raise the available XP by a specified amount
+        /// </summary>
+        /// <param name="amount">A unsigned long containing the desired XP amount to raise</param>
         public void GrantXp(ulong amount)
         {
             character.GrantXp(amount);
+            //TODO: Ask for an Answer: Does CheckForLevelup() go before or after the xp message?
+            CheckForLevelup();
             var xpTotalUpdate = new GameMessagePrivateUpdatePropertyInt64(Session, PropertyInt64.TotalExperience, character.TotalExperience);
             var xpAvailUpdate = new GameMessagePrivateUpdatePropertyInt64(Session, PropertyInt64.AvailableExperience, character.AvailableExperience);
             var message = new GameMessageSystemChat($"{amount} experience granted.", ChatMessageType.Broadcast);
             Session.WorldSession.EnqueueSend(xpTotalUpdate, xpAvailUpdate, message);
         }
 
+        /// <summary>
+        /// Determines if the player has advanced a level
+        /// </summary>
+        /// <remarks>
+        /// Known issues:
+        ///         1. Have not tested up to max level, so there isn't any special text or fireworks for max level yet.
+        /// </remarks>
         private void CheckForLevelup()
         {
+            // Question: Where do *we* call CheckForLevelup()? : 
+            //      From within the player.cs file, the options might be:
+            //           GrantXp()
+            //      From outside of the player.cs file, we may call CheckForLevelup() durring? :
+            //           XP Updates?
+
             var chart = DatabaseManager.Charts.GetLevelingXpChart();

```

Comment:
In this function, there is now code that will check the expenditure charts against the current total experience. Since experience is cumulative, when the total amount of experience gained has exceeded the total experience for a rank level on the chart, then the function will advance to the next Rank in. The object returned from the next rank of the chart will determine if the next level is reached.